src/networks/infinitesquarenetwork.jl

```diff
@@ -74,42 +91,60 @@ function Base.isapprox(A₁::InfiniteSquareNetwork, A₂::InfiniteSquareNetwork;
 end
 
 ## Rotations
-function Base.rotl90(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,2}} # Rotations of matrix unit cells
-    return NWType(rotl90(rotl90.(unitcell(A))))
-end
-function Base.rotr90(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,2}}
-    return NWType(rotr90(rotr90.(unitcell(A))))
-end
-function Base.rot180(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,2}}
-    return NWType(rot180(rot180.(unitcell(A))))
+function Base.rotl90(n::InfiniteSquareNetwork)
+    return InfiniteSquareNetwork(rotl90(rotl90.(unitcell(n))))
 end
-function Base.rotl90(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,3}} # Rotations of cubic unit cells along z-axis
-    return NWType(stack(rotl90, eachslice(unitcell(A); dims=3)))
+function Base.rotr90(n::InfiniteSquareNetwork)
+    return InfiniteSquareNetwork(rotr90(rotr90.(unitcell(n))))
 end
-function Base.rotr90(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,3}}
-    return NWType(stack(rotr90, eachslice(unitcell(A); dims=3)))
+function Base.rot180(n::InfiniteSquareNetwork)
+    return InfiniteSquareNetwork(rot180(rot180.(unitcell(n))))
 end
-function Base.rot180(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,3}}
-    return NWType(stack(rot180, eachslice(unitcell(A); dims=3)))
+
+## Chainrules
+function ChainRulesCore.rrule(
+    ::typeof(Base.getindex), network::InfiniteSquareNetwork, args...
+)
+    O = network[args...]
+
+    function getindex_pullback(ΔO_)
+        ΔO = unthunk(ΔO_)
+        if ΔO isa Tangent # TODO: figure out why this happens in the first place...
+            ΔO = ChainRulesCore.construct(typeof(O), ChainRulesCore.backing(ΔO))
+        end
```

Comment:
I have no idea why, but this always gets a `Tangent` of the expected local sandwich type instead of a normal instance of the local sandwich type. Everywhere I check with `@showtypeofgrad` the adjoint being accumulated has the expected type, but still this somehow gets a `Tangent`.

The explicit `construct` fixes things so this is probably fine, but it would still be nice to not have to do this...

Reply:
I updated things to not use any custom rrules for component getters like `ket` and `bra`, and I can make things work by properly unthunking components and then converting the tangents here. I think this is cleaner than hooking into all the getters to make sure there's no tangent types in the first place, so probably things are just fine like this.